chalk-ir/src/lib.rs

```diff
@@ -747,6 +747,7 @@ impl InferenceVar {
 #[derive(Clone, PartialEq, Eq, Hash, HasInterner)]
 pub struct Fn<I: Interner> {
     pub num_binders: usize,
+    pub abi: I::FnAbi,
```

Comment:
We need to be checking in type equality that the ABIs are the same. I guess that, given how chalk is currently setup, this would take place here

https://github.com/rust-lang/chalk/blob/4dba5a5c13c6a96375a889700c9bfc6a5f47b577/chalk-solve/src/infer/unify.rs#L146-L148

i.e., just before calling `unify_binders`, but I don't love that this slipped through. I think the *best* place for it to be checked would be in the `zip` code, i.e., there should be a `super_zip_tys` function or something that manages some of the grungy details, but I definitely think that's out of scope for this PR, so no need to consider anything like that.

We should add a test for unifying function types with distinct ABIs, basically testing that this goal gets an error

```
extern "C" fn() == extern "Rust" fn()
```